### PR TITLE
feat: add subscriptions

### PR DESCRIPTION
**Pull request by sdanialraza**
**Please describe the changes this PR makes and why it should be merged:**
Adds subscriptions for mainlib

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

**Review by OfficialSirH — APPROVED** [review]
Me parece bien

File: packages/discord.js/src/structures/Subscription.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+       * @type {?number}
+       */
+      this.canceledTimestamp = data.canceled_at ? Date.parse(data.canceled_at) : null;
+    } else {
+      this.canceledTimestamp ??= null;
+    }
+
+    if ('country' in data) {
+      /**
+       * ISO3166-1 alpha-2 country code of the payment source used to purchase the subscription.
```
Comment (Jiralite) [inline packages/discord.js/src/structures/Subscription.js:71]:
```suggestion
       * ISO 3166-1 alpha-2 country code of the payment source used to purchase the subscription.
```

File: packages/discord.js/src/managers/SubscriptionManager.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+  /**
+   * Fetches subscriptions for this application
+   * @param {FetchSubscriptionOptions|FetchSubscriptionsOptions} [options={}] Options for fetching the subscriptions
+   * @returns {Promise<Collection<Snowflake, Subscription>>}
+   */
+  async fetch(options = {}) {
+    if (typeof options !== 'object') throw new DiscordjsTypeError(ErrorCodes.InvalidType, 'options', 'object', true);
+
+    const { after, before, cache = true, limit = 50, sku, subscriptionId, user } = options;
```
Comment (Jiralite) [inline packages/discord.js/src/managers/SubscriptionManager.js:52]: We don't seem to hard-code a limit elsewhere. How come we do for this?

Comment (sdanialraza) [reply packages/discord.js/src/managers/SubscriptionManager.js:52]: Good question

File: packages/discord.js/src/index.js [hunk 3]
```diff
@@ -83,6 +83,7 @@ exports.ReactionManager = require('./managers/ReactionManager');
 exports.ReactionUserManager = require('./managers/ReactionUserManager');
 exports.RoleManager = require('./managers/RoleManager');
 exports.StageInstanceManager = require('./managers/StageInstanceManager');
+exports.SubscriptionManager = require('./managers/SubscriptionManager');
```
Comment (vladfrangu) [inline packages/discord.js/src/index.js:86]: You're using `exports.<class>` syntax in the file, you need to either deconstruct the class name from the returned value from require (`require().class`), or move to `module.exports`

File: packages/discord.js/src/index.js [hunk 4]
```diff
@@ -193,6 +194,7 @@ exports.SKU = require('./structures/SKU').SKU;
 exports.StringSelectMenuOptionBuilder = require('./structures/StringSelectMenuOptionBuilder');
 exports.StageChannel = require('./structures/StageChannel');
 exports.StageInstance = require('./structures/StageInstance').StageInstance;
+exports.Subscription = require('./structures/Subscription');
```
Comment (vladfrangu) [inline packages/discord.js/src/index.js:197]: Ditto

**Review by vladfrangu — DISMISSED** [review]

**Review by Jiralite — DISMISSED** [review]
According to the documentation[^1] and the specification[^2], `ends_at` may now be `null`. This means this logic needs to be modified as `Date.parse(null)` will result in `NaN`:

https://github.com/discordjs/discord.js/blob/f2f7f1f65b29149ac04db60e624a600036277810/packages/discord.js/src/structures/Entitlement.js#L84-L93

I say this, but the [entitlement object](https://discord.com/developers/docs/resources/entitlement#entitlement-object) itself on the documentation does not have it marked as nullable. It probably needs updating (https://github.com/discord/discord-api-docs/issues/7268).

[^1]: https://discord.com/developers/docs/monetization/implementing-app-subscriptions#how-app-subscriptions-work
[^2]: https://github.com/discord/discord-api-spec/blob/c979cb269dbaf8cd09b8d0a39cc683d9623d8ce4/specs/openapi.json#L16491-L16495

File: packages/discord.js/typings/index.d.ts [hunk 5]
```diff
@@ -4094,6 +4112,26 @@ export class EntitlementManager extends CachedManager<Snowflake, Entitlement, En
   public consume(entitlementId: Snowflake): Promise<void>;
 }
 
+export interface FetchSubscriptionOptions extends BaseFetchOptions {
+  cache?: boolean;
+  sku: SKUResolvable;
+  subscriptionId: Snowflake;
+}
```
Comment (Jiralite) [inline packages/discord.js/typings/index.d.ts:4119]: `BaseFetchOptions` already has `cache`:

https://github.com/discordjs/discord.js/blob/3669d5e1120ed0a6e8d8cb8c48414d7841efd594/packages/discord.js/typings/index.d.ts#L4964-L4967

```suggestion
export interface FetchSubscriptionOptions extends BaseFetchOptions {
  sku: SKUResolvable;
  subscriptionId: Snowflake;
}
```

**Review by Jiralite — CHANGES_REQUESTED** [review]

**Review by Jiralite — APPROVED** [review]

File: packages/discord.js/src/client/websocket/handlers/SUBSCRIPTION_DELETE.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
+'use strict';
+
+const Events = require('../../../util/Events');
+
+module.exports = (client, { d: data }) => {
+  const subscription = client.application.subscriptions.cache._add(data, false);
```
Comment (Qjuh) [inline packages/discord.js/src/client/websocket/handlers/SUBSCRIPTION_DELETE.js:6]:
```suggestion
  const subscription = client.application.subscriptions._add(data, false);
```

**Review by Qjuh — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/structures/Subscription.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+     * The start of the current subscription period
+     * @type {Date}
+     */
+    this.currentPeriodStart = new Date(data.current_period_start);
+
+    /**
+     * The end of the current subscription period
+     * @type {Date}
+     */
+    this.currentPeriodEnd = new Date(data.current_period_end);
```
Comment (vladfrangu) [inline packages/discord.js/src/structures/Subscription.js:51]: Shouldn't this follow the same structure as the other timestamp fields?

**Review by vladfrangu — APPROVED** [review]

**Review by didinele — APPROVED** [review]

**Pull request merged**